### ISSUE TITLE
BatchNorm2d return different values with the same inputs

### ISSUE THREAD

**Issue by T1anZhenYu** [pytorch/pytorch#41463]
## 🐛 Bug

I created  Class"A" with a BN layer and 2 conv layers and only BN layer was uesd in "A.forward()". I also created "NetA" with 2 "A" layers and a conv layer , but only one "A" layer and the conv layer were uesd in  "NetA.forward()".  I also created "NetB" which was the same as NetA but the unnecessary "A" layer was deleted. I trained NetA for 20 epochs and test NetA and the results were recoreded as "rA".  NetB was trained in the same manner as NetA and the result were recoreded as "rB". rA should be the same with rB. But I found they are different. 

## To Reproduce

Class A
```python
import torch 
import torch.nn as nn 
import torch.optim as optim
import math
torch.manual_seed(1)
class A(nn.Module):
    def __init__(self, num_features, eps=1e-05, momentum=0.9, affine=True):
        super(A, self).__init__()
        ks = 5
        self.a = nn.Conv1d(1, 1, kernel_size=ks, padding=(ks-1) // 2, bias=False) 
        self.b = nn.Conv1d(1, 1, kernel_size=ks, padding=(ks-1) // 2, bias=False) 
        self.bn = nn.BatchNorm2d(num_features,affine = False)
    def forward(self, x):
        out = self.bn(x)
        return out
```
NetA
```python
class NetA(nn.Module):
  def __init__(self):
    super(NetA, self).__init__()

    self.input = torch.nn.parameter.Parameter(100*torch.arange(50.).reshape(1,2,5,5))

    self.bn1 = A(2,affine=False)
    self.bn2 = A(2,affine=False)
    self.conv1 = nn.Conv2d(2,2,2,1,0)
    self.conv2 = nn.Conv2d(2,2,2,1,0)
    torch.nn.init.constant_(self.conv1.weight, 2)
    torch.nn.init.constant_(self.conv2.weight,2)
  def forward(self):
    x = self.input
    x = self.bn1(x)
    x = self.conv1(x)
```
NetB
```python
class NetB(nn.Module):
  def __init__(self):
    super(NetB, self).__init__()

    self.input = torch.nn.parameter.Parameter(100*torch.arange(50.).reshape(1,2,5,5))

    self.bn1 = A(2,affine=False)
    # self.bn2 = A(2,affine=False)
    self.conv1 = nn.Conv2d(2,2,2,1,0)
    self.conv2 = nn.Conv2d(2,2,2,1,0)
    torch.nn.init.constant_(self.conv1.weight, 2)
    torch.nn.init.constant_(self.conv2.weight,2)
  def forward(self):
    x = self.input
    x = self.bn1(x)
    x = self.conv1(x)
```
Train and test 
```python 
model = NetA()#model = NetB() if you want to test NetB()

model.train()
# label = torch.reshape(torch.range(1,32.),[1,2,4,4])
optimizer = optim.SGD(model.parameters(), lr=1,)
print('input')
print(model.input[0,0,0,:])


for i in range(20):
  y = model()
  loss = torch.sum(y)
  optimizer.zero_grad()
  loss.backward()
  optimizer.step()

model.eval()
y = model()
print(model)
print('test')
print(y[0,0,0,:])
```

## Expected behavior
NetA() and NetB() 's log should be the same.

```

 - PyTorch Version (1.5.1):
 - OS ( Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:3.6.9
 - CUDA/cuDNN version:cu101
 - GPU models and configuration:
 - Any other relevant information:

**Comment by ezyang** [comment]
Batchnorms internally maintain statistics which will change over time. This is expected behavior.

**Comment by T1anZhenYu** [comment]
But I don't understand the relationship between Batchnorm layer initialization and statistics.  Can you explain it? thanks

**Comment by T1anZhenYu** [comment]
I got a different result when I remove "self.bn2" initialization from NetA, even though "self.bn2" is not used in NetA.forward().  Is this expected?

**Comment by ezyang** [comment]
Oh, I misinterpreted your example. But I still think it is expected. You are probably perturbing RNG state in some way. You can use the generator api to explicitly save and restore rng state before and after the bn2 call, and that hsould make the divergence go away.
